src/v/config/property.h

```diff
@@ -568,6 +597,70 @@ class conversion_binding : public binding_base<T> {
     }
 };
 
+/**
+ * A property sanctioning binding contains a copy of the property's
+ * value, which will be updated in-place whenever the property is
+ * updated in the cluster configuration. It offers both free access
+ * to the value and a sanctioned view, in case the value is restricted.
+ *
+ * This is useful for classes that want a copy of a property without
+ * having to write their own logic for subscribing to value changes.
+ */
+template<class T>
+class sanctioning_binding : public binding_base<T> {
+private:
+    T _value;
+    T _default_value;
+    bool _is_value_restricted;
```

Comment:
If you are asking about the bindings in general, I... don't know.

If you are asking about this specific binding, we need all three, as we don't have access in the property handle in the binding itself. Only the `binding_base` has a private handle and all the communication happens through the `do_update` function.

Reply:
I was talking about the `is_restricted` specifically, but yeah I misread the `parent()` call in the constructor as an accessor.

I think I understand why a binding is useful (even w/o watchers), but I guess since `check_restricted` is cheap and not stateful, it seems like we could query the parent property live (e.g. by adding `base_property::check_restricted()` and having that call into the parent with the current value).

again, maybe it's fine, just trying to zero in on what's strictly required in terms of additional state & API.

Reply:
I am conflicted on this.

On one hand, I don't want to expose a check that circumvents the cached values and goes straight to the parent, cause it feels against the logic of the binding. While the `check_restricted` that we currently have are indeed cheap, it could be a generic functor or a big vector, so in theory the cost is unbounded.

On the other hand, having a `check_restricted` call exposed on the binding would allow for ~no~ less duplicate logic. Specifically with the leadership preference. But it is already duplicated in a few places, already.

For now, the "all cached and no `check_restricted` on the binding" is where I lean.

Reply:
> What is the original reasoning for never ever calling up to the parent property? Does that reasoning extend to querying data that never ever change?

I have been trying to understand the same thing.
The only theoretical scenario I can come up with that it would be dangerous is if the `config::property` get's destroyed before the binding. The current semantics allow for it and the binding is left with the last value and a `nullptr` `_parent`. Noone is accessing the `_parent` from within `binding_base` apart from the constructors (and `detach` which is called on the `~property`.
However, given the context of these objects, i doubt this is a scenario we should encounter.

> If we relax (1), does the diff become significantly smaller

not significantly, I don't think so.
It is worth noting that this PR is in essence 4 different ones. One for the `sanction_binding` and one for each of the three sanctions. It did grow quite a bit :sweat_smile: